### PR TITLE
Convert `Peerset` into a shared sync struct

### PR DESCRIPTION
**Pull request by dmitry-markin**
Resolves https://github.com/paritytech/substrate/issues/11922.

### Follow-up

Things could be simplified further by removing `Message::Accept` and `Message::Reject` variants from [`sc_peerset::Message`](https://github.com/paritytech/substrate/blob/e055af136912a7b21ecd619d782c81e3281b8e58/client/peerset/src/lib.rs#L202-L225) and making [`Peerset::incoming()`](https://github.com/paritytech/substrate/blob/e055af136912a7b21ecd619d782c81e3281b8e58/client/peerset/src/lib.rs#L172-L176) return result synchronously. Unfortunately, this requires modifying `Notifications`, and as agreed with @altonen and @bkchr should be done only after it's better covered by tests.

### CODE REVIEW

File: client/peerset/src/lib.rs [hunk 1]
```diff
@@ -131,50 +125,75 @@ impl PeersetHandle {
...
 	pub fn remove_from_peers_set(&self, set_id: SetId, peer_id: PeerId) {
-		let _ = self.tx.unbounded_send(Action::RemoveFromPeersSet(set_id, peer_id));
+		self.inner.lock().remove_from_peers_set(set_id, peer_id);
 	}
 
 	/// Returns the reputation value of the peer.
-	pub async fn peer_reputation(self, peer_id: PeerId) -> Result<i32, ()> {
-		let (tx, rx) = oneshot::channel();
+	pub fn peer_reputation(&self, peer_id: PeerId) -> impl Future<Output = Result<i32, ()>> + Send {
+		// TODO: refactor into sync function.
```
Comment (dmitry-markin) [inline client/peerset/src/lib.rs:166]: Missed this one, converting into draft.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2198836

File: client/peerset/src/lib.rs [hunk 2]
```diff
@@ -714,52 +712,27 @@ impl Peerset {
...
-					self.alloc_slots(SetId(set_index));
-				}
-			}
-
-			let action = match Stream::poll_next(Pin::new(&mut self.rx), cx) {
-				Poll::Pending => return Poll::Pending,
-				Poll::Ready(Some(event)) => event,
-				Poll::Ready(None) => return Poll::Pending,
-			};
+	fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
```
Comment (altonen) [inline client/peerset/src/lib.rs:723]: I'm not a huge fan of this design choice. Previously there was a clear difference between the protocols that would use `Peerset` and the protocol that owned `Peerset` and I think it made sense. Now it seems that any protocol that acquires a handle to `Peerset` can poll it and drain events from it even if it has no idea what to do with them.

Comment (dmitry-markin) [reply client/peerset/src/lib.rs:723]: Would splitting `Peerset` into `Peerset` (lacking polling) and `PeersetStream`, both actually containing an `Arc` make it better?

Comment (dmitry-markin) [reply client/peerset/src/lib.rs:723]: The naming can be different, of course.

Comment (altonen) [reply client/peerset/src/lib.rs:723]: I think splitting them into two would be good.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2202580

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2202581

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2202565

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2202882

**Comment by dmitry-markin** [comment]
A note to not forget what is happening here. Before this PR, `Peerset` had a channel for incoming commands sent via `PeersetHandle` (`rx`). These commands were only processes in `poll()` after all the outgoing events  from `message_queue` with commands for `Notifications` were sent out. This way `Notifications` received consistent stream of events from `Peerset`.

After removing the channel for incoming commands and making the interface for incoming commands synchronous, it became possible that the command from `Notifications` and the event-reaction to it are interleaved by old out-events not yet dispatched from `message_queue`. As a result, the interface contract between `Notifications` and `Peerset` is broken.

One possible solution would be to make the `Notifications`-looking part of the `Peerset` interface also synchronous. E.g., instead of pushing the out-event with `Accept` and `Reject` to the end of the queue, we can return the decision immediately from `Peerset::incoming()`.  This requires changing the `Notifications` code and is blocked by https://github.com/paritytech/substrate/pull/13033.

Nevertheless, it's unclear whether this would be sufficient, because `Connect` and `Drop` out-events would still be buffered in the `message_queue` and only delivered to `Notifications` during polling.

**Comment by dmitry-markin** [comment]
Yes, I'm going to continue working on this.

**Comment by altonen** [comment]
> Yes, I'm going to continue working on this.

Btw, based on [this comment](https://github.com/paritytech/substrate/issues/5685#issuecomment-1327236882), there may be some quite significant changes to the peerset at some point.

**Comment by dmitry-markin** [comment]
> Btw, based on [this comment](https://github.com/paritytech/substrate/issues/5685#issuecomment-1327236882), there may be some quite significant changes to the peerset at some point.

Yeah, the race conditions @tomaka is mentioning are also in the `Peerset`, and exactly the reason why this refactoring was not finished in one go. May be if we manage to make `Peerset` synchronous, it would be easier to transit to synchronous outer (external) API. Or we can try to crystallize the solution with outer API and skip the step of making `Peerset` synchronous completely.

**Comment by dmitry-markin** [comment]
Superseded by https://github.com/paritytech/substrate/pull/13611. Closing this PR.

**Pull request closed without merging**